### PR TITLE
Mnist example : Assets

### PR DESCRIPTION
**Pull request by Fabien-GELUS**

### CODE REVIEW

File: examples/mnist/assets/objective/description.md [hunk 1]
```diff
@@ -0,0 +1 @@
+# Mnist : hand-drawn digit recognising
```
Comment (natct10) [inline examples/mnist/assets/objective/description.md]: No space before ":"

**Review by natct10 — COMMENTED** [review]
Looks nice, well done @Fabien-GELUS 
I'll test it!
cc @bowni

**Comment by Fabien-GELUS** [comment]
You can test the scripts now, see the readme for the commands

File: examples/mnist/assets/objective/description.md (hunk 1 above)
Comment (Fabien-GELUS) [reply examples/mnist/assets/objective/description.md]: https://github.com/Fabien-GELUS/substra/pull/3/commits/e56cb7869ccc0cfc27bc9e6de5f2ff18caa20d07

File: examples/mnist/scripts/add_dataset_objective.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        pg.close()
+        substra_logger.removeHandler(progress_handler)
+        substra_logger.addHandler(default_stream_handler)
+
+
+current_directory = os.path.dirname(__file__)
+assets_directory = os.path.join(current_directory, '../assets')
+
+client = substra.Client(profile_name="node-1")
+client.login()
```
Comment (natct10) [inline examples/mnist/scripts/add_dataset_objective.py]: You can remove this `client.login()`, it has been recently removed: https://github.com/SubstraFoundation/substra/commit/aee8d45cbdab99c285f1dccf7dca109595de6f99

**Review by natct10 — COMMENTED** [review]
You can now remove the `client.login()`, see: https://github.com/SubstraFoundation/substra/commit/aee8d45cbdab99c285f1dccf7dca109595de6f99

File: examples/mnist/scripts/add_dataset_objective.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+DATASET = {
+    'name': 'Mnist',
+    'type': 'npy',
+    'data_opener': os.path.join(assets_directory, 'dataset/opener.py'),
+    'description': os.path.join(assets_directory, 'dataset/description.md'),
```
Comment (natct10) [inline examples/mnist/scripts/add_dataset_objective.py]: I am getting a permissions issues (`{'permissions': ['This field may not be null.']}`), can you please update as it was done here: https://github.com/SubstraFoundation/substra/pull/168

**Review by natct10 — APPROVED** [review]

**Review by natct10 — CHANGES_REQUESTED** [review]
Update permissions & remove login

File: examples/mnist/scripts/add_train_algo_cnn.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+import os
+import zipfile
+
+import substra
+
+current_directory = os.path.dirname(__file__)
+assets_directory = os.path.join(current_directory, '../assets')
+
+client = substra.Client(profile_name="node-1")
+client.login()
```
Comment (natct10) [inline examples/mnist/scripts/add_train_algo_cnn.py]: Can you remove this one too, please?

**Review by natct10 — CHANGES_REQUESTED** [review]
One remaining .login()

File: examples/mnist/assets/algo_cnn/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# this base image works in both CPU and GPU enabled environments
+FROM substrafoundation/substra-tools:0.5.0
+
+# install dependencies
+RUN pip3 install keras tensorflow
```
Comment (RomainGoussault) [inline examples/mnist/assets/algo_cnn/Dockerfile]: I don't have much experience with Dockerfile, but can you specify the keras/tensorflow version ?

File: examples/mnist/assets/algo_cnn/algo.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        return X
+
+
+    def train(self, X, y, models, rank):
+        X = self._normalize_X(X)
+        num_classes = 10
+        y = keras.utils.to_categorical(y, num_classes)
+        
+
+
```
Comment (RomainGoussault) [inline examples/mnist/assets/algo_cnn/algo.py]: lots of empty space.
You could use a formatter like https://github.com/psf/black
cc @natct10

File: examples/mnist/assets/dataset/opener.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+class MnistOpener(tools.Opener):
+
+    @classmethod
+    def _get_files(cls, folders):
+        """Return list of X and y file given a folder location"""
+        X_files, y_files = [], []
+
+        for folder in folders:
+
+            Xs = glob.glob(os.path.join(folder, 'x*.npy'))
```
Comment (RomainGoussault) [inline examples/mnist/assets/dataset/opener.py]: `Xs`, what it is the meaning of the `s` ?

Comment (RomainGoussault) [reply examples/mnist/assets/dataset/opener.py]: maybe rename to something like `X_current_folder`

File: examples/mnist/assets/objective/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,11 @@
+# this base image works in both CPU and GPU enabled environments
+FROM substrafoundation/substra-tools:0.5.0
+
+# install dependencies
+RUN pip3 install sklearn
```
Comment (RomainGoussault) [inline examples/mnist/assets/objective/Dockerfile]: idem, is it possible to specify a version ?

File: examples/mnist/scripts/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,5 @@
+tqdm
+sklearn
+pandas
+keras
+tensorflow
```
Comment (RomainGoussault) [inline examples/mnist/scripts/requirements.txt]: add version

**Review by RomainGoussault — COMMENTED** [review]
Super work !! 👍

File: examples/mnist/assets/algo_cnn/Dockerfile (hunk 5 above)
Comment (natct10) [reply examples/mnist/assets/algo_cnn/Dockerfile]: Yep on doit pouvoir lui passer un `requirements.txt`

File: examples/mnist/assets/algo_cnn/algo.py (hunk 6 above)
Comment (natct10) [reply examples/mnist/assets/algo_cnn/algo.py]: <3

File: examples/mnist/assets/algo_cnn/algo.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        else:
+            X = X.reshape(X.shape[0], img_rows, img_cols, 1)
+
+            input_shape = (img_rows, img_cols, 1)
+
+        X = X.astype("float32")
+
+        X /= 255
+
+        print("X shape:", X.shape)
```
Comment (natct10) [inline examples/mnist/assets/algo_cnn/algo.py]: Here, fyi, you could use fstrings => print(f"X shape: {X.shape}")

File: examples/mnist/scripts/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,5 @@
+tqdm==4.46.1
+sklearn==0.0
+pandas==1.0.4
+Keras==2.3.1
+tensorflow==2.2.0
```
Comment (natct10) [inline examples/mnist/scripts/requirements.txt:5]: :+1:

File: examples/mnist/assets/objective/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,11 @@
+# this base image works in both CPU and GPU enabled environments
+FROM substrafoundation/substra-tools:0.5.0
+
+# install dependencies
+RUN pip3 install sklearn==0.0
```
Comment (RomainGoussault) [inline examples/mnist/assets/objective/Dockerfile]: 0.0 ?

File: examples/mnist/scripts/generate_data_samples.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
+# Copyright 2018 Owkin, inc.
```
Comment (RomainGoussault) [inline examples/mnist/scripts/generate_data_samples.py]: To remove ? (same for the other files)

File: examples/mnist/assets/objective/Dockerfile (hunk 12 above)
Comment (natct10) [reply examples/mnist/assets/objective/Dockerfile]: il faudrait copier le requirements.txt et `pip3 install -r requirements.txt`

Comment (natct10) [reply examples/mnist/assets/objective/Dockerfile]: Oh I got the same _version_: 
```sh
pip3 list | grep sklearn
sklearn  0.0
```

**Comment by Fabien-GELUS** [comment]
The scripts `add_dataset_objective.py `and `add_train_algo_cnn.py `have been tested on substra (cli 0.5.0, hlf-k8s 0.0.12 , backend 0.0.14) and ran successfully ! You can now train and test the dataset one a single node with substra. 
```
substra get testtuple 73238df59642d645c7929703a275e18dae48c2a0759dfc61e8ed0e086f541142
KEY                         73238df59642d645c7929703a275e18dae48c2a0759dfc61e8ed0e086f541142
TRAINTUPLE KEY              91763e067c0ee7ba6458c82c32d9cd8bc7c46cdd5701079eb6c763ac4d3fac6e
TRAINTUPLE TYPE             traintuple
ALGO KEY                    c39addcc4a23a57ddde1e68bad76991bbe991648b2720e3bfe4054296796a9ff
ALGO NAME                   Mnist: CNN
OBJECTIVE KEY               683a4ad0c8f8169d832c669230a0b4fd43dadcd00daf363ba42dae116e46d2b4
CERTIFIED                   True
STATUS                      done
PERF                        0.9797
DATASET KEY                 454711d6bcee7ab4aba78192ba2c521a429c100d532534d3e89765c4799d7068
TEST DATA SAMPLE KEYS       2 keys
RANK                        0
TAG                         
COMPUTE PLAN ID             
LOG                         
CREATOR                     MyOrg1MSP
WORKER                      MyOrg1MSP
PERMISSIONS                 Processable by its owner only
```

File: examples/mnist/assets/algo_cnn/algo.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        X /= 255
+
+        print(f"X shape: {X.shape}")
+        print(f"{X.shape[0]} samples")
+
+        return X
+
+    def train(self, X, y, models, rank):
+        X = self._normalize_X(X)
+        num_classes = 10
```
Comment (bowni) [inline examples/mnist/assets/algo_cnn/algo.py]: Couldn't we get that from `y` instead?

File: examples/mnist/assets/algo_cnn/algo.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+        model.add(Flatten())
+        model.add(Dense(128, activation="relu"))
+        model.add(Dropout(0.5))
+        model.add(Dense(num_classes, activation="softmax"))
+
+        model.compile(
+            loss=keras.losses.categorical_crossentropy,
+            optimizer=keras.optimizers.Adadelta(),
+            metrics=["accuracy"],
+        )
```
Comment (bowni) [inline examples/mnist/assets/algo_cnn/algo.py:53]: You could also put that in a dedicated `initialize_new_model()` function to keep your `train()` function shorter / easier to read

File: examples/mnist/assets/objective/metrics.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+import substratools as tools
+
+
+class MnistMetrics(tools.Metrics):
+    def score(self, y_true, y_pred):
+        """Returns the macro-average recall
+
+        :param y_true: actual values from test data
+        :type y_true: pd.DataFrame
+        :param y_true: predicted values from test data
```
Comment (bowni) [inline examples/mnist/assets/objective/metrics.py]: Typo `true` -> `pred`

File: examples/mnist/assets/algo_cnn/algo.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+        epochs = 1
+        batch_size = 128
```
Comment (bowni) [inline examples/mnist/assets/algo_cnn/algo.py]: I don't have the whole context in mind, but why just `1` epoch, and why those 2 iteration-related parameters hard-coded here?

File: examples/mnist/assets/algo_cnn/algo.py (hunk 14 above)
Comment (Fabien-GELUS) [reply examples/mnist/assets/algo_cnn/algo.py]: I think `y` is a list of Integer and not a list of vector (that's why we use `to_categorical` to transform it into a vector), so maybe we could try to get the max on `y` but i think it's overkill. We will always have 10 digits on MNIST.

File: examples/mnist/assets/algo_cnn/algo.py (hunk 17 above)
Comment (Fabien-GELUS) [reply examples/mnist/assets/algo_cnn/algo.py]: As it's an example and it already takes a lot of time to train the model one time on all the dataset, I chose to do only 1 epoch. I could put these parameters directly in the function `fit` maybe ?

**Review by natct10 — APPROVED** [review]

**Comment by Fabien-GELUS** [comment]
Rebase problem : reopened a new PR from a new branch

**Pull request closed without merging**